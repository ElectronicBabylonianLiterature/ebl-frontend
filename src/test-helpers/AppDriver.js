import React from 'react'
import { fireEvent, render, waitForElement } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import App from 'App'
import Auth from 'auth/Auth'
import WordRepository from 'dictionary/WordRepository'
import FragmentRepository from 'fragmentarium/FragmentRepository'
import ImageRepository from 'fragmentarium/ImageRepository'
import FragmentService from 'fragmentarium/FragmentService'
import WordService from 'dictionary/WordService'
import TextService from 'corpus/TextService'
import Session from 'auth/Session'
import BibliographyRepository from 'bibliography/BibliographyRepository'
import BibliographyService from 'bibliography/BibliographyService'
import { defaultErrorReporter } from 'ErrorReporterContext'
import createAuth0Config from 'auth/createAuth0Config'

function createApp(api, sessionStore) {
  const auth0Config = createAuth0Config()
  const auth = new Auth(sessionStore, defaultErrorReporter, auth0Config)
  const wordRepository = new WordRepository(api)
  const fragmentRepository = new FragmentRepository(api)
  const imageRepository = new ImageRepository(api)
  const bibliographyRepository = new BibliographyRepository(api)
  const wordService = new WordService(wordRepository)
  const bibliographyService = new BibliographyService(bibliographyRepository)
  const fragmentService = new FragmentService(
    fragmentRepository,
    imageRepository,
    bibliographyService
  )
  const textService = new TextService(api)

  return (
    <App
      auth={auth}
      wordService={wordService}
      fragmentService={fragmentService}
      bibliographyService={bibliographyService}
      textService={textService}
    />
  )
}

export default class AppDriver {
  initialEntries = []
  element = null
  session = null

  constructor(api) {
    this.api = api
  }

  withPath(path) {
    this.initialEntries = [path]
    return this
  }

  withSession() {
    this.session = new Session(
      'accessToken',
      'idToken',
      Number.MAX_SAFE_INTEGER,
      ['write:texts']
    )
    return this
  }

  render() {
    const fakeSessionStore = {
      setSession: session => {
        this.session = session
      },
      clearSession: () => {
        this.session = null
      },
      getSession: () => this.session || new Session('', '', 0, [])
    }

    this.element = render(
      <MemoryRouter initialEntries={this.initialEntries}>
        {createApp(this.api, fakeSessionStore)}
      </MemoryRouter>
    )

    return this
  }

  async waitForText(text) {
    await waitForElement(() => this.element.getByText(text))
  }

  expectTextContent(text) {
    expect(this.element.container).toHaveTextContent(text)
  }

  expectNotInContent(text) {
    expect(this.element.queryByText(text)).toBeNull()
  }

  expectLink(text, expectedHref) {
    expect(this.element.getByText(text)).toHaveAttribute('href', expectedHref)
  }

  expectBreadcrumbs(crumbs) {
    this.expectTextContent(crumbs.join(''))
  }

  expectInputElement(label, expectedValue) {
    expect(this.element.getByLabelText(label).value).toEqual(
      String(expectedValue)
    )
  }

  changeValueByLabel(label, newValue) {
    const input = this.element.getByLabelText(label)
    fireEvent.change(input, { target: { value: newValue } })
  }

  click(text, n = 0) {
    const clickable = this.element.getAllByText(text)[n]
    fireEvent.click(clickable)
    // await wait()
  }
}
