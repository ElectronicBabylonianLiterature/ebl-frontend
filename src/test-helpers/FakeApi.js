import Promise from 'bluebird'

export default class FakeApi {
  gets = []
  client = {
    fetchJson: (path, authenticate) => {
      const expectation = this.gets.find(entry => entry.path === path && entry.authenticate === authenticate)
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(new Error(`Unexpected ${authenticate ? 'authenticated' : 'not-authenticated'} fetchJson: ${path}`))
    },

    fetchBlob: (path, authenticate) => {
      return Promise.reject(new Error(`Unexpected ${authenticate ? 'authenticated' : 'not-authenticated'} fetchBlob: ${path}`))
    }
  }

  expectText (text) {
    this.gets.push({
      path: `/texts/${text.category}/${text.index}`,
      authenticate: true,
      response: text
    })
    return this
  }

  expectStatistics (statistics) {
    this.gets.push({
      path: `/statistics`,
      authenticate: false,
      response: statistics
    })
    return this
  }
}
