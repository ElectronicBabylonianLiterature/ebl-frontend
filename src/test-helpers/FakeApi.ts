import Promise from 'bluebird'
import Word from 'dictionary/domain/Word'

class Expectation {
  method: 'POST' | 'GET' = 'GET'
  path = ''
  authenticate = true
  response = {}
  verify = false
  called = false
  body: object | null = null
  isBlob = false

  constructor(data) {
    Object.assign(this, data)
  }
}

export default class FakeApi {
  private expectations: Expectation[] = []

  readonly client = {
    fetchJson: jest.fn().mockImplementation((path, authenticate) => {
      const expectation = this.expectations.find(
        entry =>
          entry.method === 'GET' &&
          entry.path === path &&
          entry.authenticate === authenticate &&
          !entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(
            new Error(
              `Unexpected ${
                authenticate ? 'authenticated' : 'not-authenticated'
              } fetchJson: ${path}`
            )
          )
    }),

    postJson: jest.fn().mockImplementation(path => {
      const expectation = this.expectations.find(
        entry => entry.method === 'POST' && entry.path === path && !entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(new Error(`Unexpected postJson: ${path}`))
    }),

    fetchBlob: jest.fn().mockImplementation((path, authenticate) => {
      const expectation = this.expectations.find(
        entry =>
          entry.method === 'GET' &&
          entry.path === path &&
          entry.authenticate === authenticate &&
          entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(
            new Error(
              `Unexpected ${
                authenticate ? 'authenticated' : 'not-authenticated'
              } fetchBlob: ${path}`
            )
          )
    })
  }

  expectTexts(texts): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts`,
        authenticate: false,
        response: texts,
        verify: true
      })
    )
    return this
  }

  allowText(text): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts/${text.category}/${text.index}`,
        authenticate: true,
        response: text
      })
    )
    return this
  }

  expectText(text): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts/${text.category}/${text.index}`,
        authenticate: true,
        response: text,
        verify: true
      })
    )
    return this
  }

  expectUpdateManuscripts(text, chapterIndex, manuscripts): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/texts/${text.category}/${text.index}/chapters/${chapterIndex}/manuscripts`,
        response: text,
        verify: true,
        body: manuscripts
      })
    )
    return this
  }

  expectUpdateLines(text, chapterIndex, lines): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/texts/${text.category}/${text.index}/chapters/${chapterIndex}/lines`,
        response: text,
        verify: true,
        body: lines
      })
    )
    return this
  }

  expectAnnotations(
    number: string,
    annotationDtos: readonly object[]
  ): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${number}/annotations`,
        authenticate: true,
        response: { annotations: annotationDtos },
        verify: true
      })
    )
    return this
  }

  expectUpdateAnnotations(
    number: string,
    annotationDtos: readonly object[]
  ): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/fragments/${number}/annotations`,
        body: {
          fragmentNumber: number,
          annotations: annotationDtos
        },
        authenticate: true,
        response: { annotations: annotationDtos },
        verify: true
      })
    )
    return this
  }

  expectFragment(fragmentDto): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${fragmentDto._id}`,
        authenticate: true,
        response: fragmentDto,
        verify: true
      })
    )
    return this
  }

  expectPhoto(number: string): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${number}/photo`,
        authenticate: true,
        isBlob: true,
        verify: true
      })
    )
    return this
  }

  expectSearchWords(query: string, words: readonly Word[]): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/words?query=${encodeURIComponent(query)}`,
        response: words,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  expectWord(word: Word): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/words/${encodeURIComponent(word._id)}`,
        response: word,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  expectUpdateWord(word: Word): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/words/${encodeURIComponent(word._id)}`,
        body: word,
        response: word,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  allowStatistics(statistics): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/statistics`,
        authenticate: false,
        response: statistics
      })
    )
    return this
  }

  allowImage(file): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/images/${file}`,
        authenticate: false,
        isBlob: true
      })
    )
    return this
  }

  verifyExpectations(): void {
    const methods = {
      GET: (expectation: Expectation): void =>
        expect(
          expectation.isBlob ? this.client.fetchBlob : this.client.fetchJson
        ).toHaveBeenCalledWith(expectation.path, expectation.authenticate),
      POST: (expectation: Expectation): void =>
        expect(this.client.postJson).toHaveBeenCalledWith(
          expectation.path,
          expectation.body || expect.anything()
        )
    }
    this.expectations
      .filter(expectation => expectation.verify)
      .forEach(expectation => methods[expectation.method](expectation))
  }
}
