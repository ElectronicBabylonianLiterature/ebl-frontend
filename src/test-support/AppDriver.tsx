import React from 'react'
import {
  fireEvent,
  render,
  RenderResult,
  act,
  Matcher,
  within,
  waitFor,
} from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import _ from 'lodash'
import App from 'App'
import WordRepository from 'dictionary/infrastructure/WordRepository'
import FragmentRepository from 'fragmentarium/infrastructure/FragmentRepository'
import ApiImageRepository from 'fragmentarium/infrastructure/ImageRepository'
import FragmentService from 'fragmentarium/application/FragmentService'
import WordService from 'dictionary/application/WordService'
import TextService from 'corpus/application/TextService'
import MemorySession, { Session, guestSession } from 'auth/Session'
import BibliographyRepository from 'bibliography/infrastructure/BibliographyRepository'
import BibliographyService from 'bibliography/application/BibliographyService'
import FragmentSearchService from 'fragmentarium/application/FragmentSearchService'
import { Promise } from 'bluebird'
import { submitForm } from 'test-support/utils'
import { eblNameProperty, AuthenticationContext } from 'auth/Auth'
import SignRepository from 'signs/infrastructure/SignRepository'
import SignService from 'signs/application/SignService'

function createApp(api): JSX.Element {
  const wordRepository = new WordRepository(api)
  const fragmentRepository = new FragmentRepository(api)
  const imageRepository = new ApiImageRepository(api)
  const bibliographyRepository = new BibliographyRepository(api)
  const wordService = new WordService(wordRepository)
  const bibliographyService = new BibliographyService(bibliographyRepository)
  const fragmentService = new FragmentService(
    fragmentRepository,
    imageRepository,
    wordRepository,
    bibliographyService
  )
  const fragmentSearchService = new FragmentSearchService(fragmentRepository)
  const textService = new TextService(api, fragmentService, wordService)
  const signsRepository = new SignRepository(api)
  const signService = new SignService(signsRepository)
  return (
    <App
      signService={signService}
      wordService={wordService}
      fragmentService={fragmentService}
      fragmentSearchService={fragmentSearchService}
      bibliographyService={bibliographyService}
      textService={textService}
    />
  )
}

export default class AppDriver {
  private readonly api
  private initialEntries: string[] = []
  private view: RenderResult | null = null
  private session: Session | null = null

  constructor(api) {
    this.api = api
  }

  getView(): RenderResult {
    if (this.view) {
      return this.view
    } else {
      throw new Error('getElement called before render.')
    }
  }

  withPath(path: string): AppDriver {
    this.initialEntries = [path]
    return this
  }

  withSession(): AppDriver {
    this.session = new MemorySession([
      'write:texts',
      'read:fragments',
      'annotate:fragments',
      'read:words',
    ])
    return this
  }

  async render(): Promise<AppDriver> {
    await act(async () => {
      this.view = render(
        <MemoryRouter initialEntries={this.initialEntries}>
          <AuthenticationContext.Provider
            value={{
              login: _.noop,
              logout: _.noop,
              getSession: (): Session => this.session ?? guestSession,
              isAuthenticated: (): boolean => this.session !== null,
              getAccessToken(): Promise<string> {
                throw new Error('Not implemented')
              },
              getUser(): { [eblNameProperty]: string } {
                return { [eblNameProperty]: 'Test' }
              },
            }}
          >
            {createApp(this.api)}
          </AuthenticationContext.Provider>
        </MemoryRouter>
      )
    })

    return this
  }

  async waitForText(text: Matcher): Promise<void> {
    await this.getView().findByText(text)
  }

  async waitForTextToDisappear(text: Matcher): Promise<void> {
    await waitFor(() => {
      expect(this.getView().queryByText(text)).not.toBeInTheDocument()
    })
  }

  expectTextContent(text: string | RegExp): void {
    expect(this.getView().container).toHaveTextContent(text)
  }

  expectNotInContent(text: Matcher): void {
    expect(this.getView().queryByText(text)).not.toBeInTheDocument()
  }

  expectLink(text: Matcher, expectedHref: string): void {
    expect(this.getView().getByText(text)).toHaveAttribute('href', expectedHref)
  }

  expectBreadcrumbs(crumbs: readonly string[]): void {
    this.expectTextContent(crumbs.join(''))
  }

  expectBreadcrumb(crumb: string, link: string): void {
    expect(
      within(this.getView().getByLabelText('breadcrumb')).getByText(crumb)
    ).toHaveAttribute('href', link)
  }

  expectInputElement(label: Matcher, expectedValue: unknown): void {
    expect(this.getView().getByLabelText(label)).toHaveValue(
      String(expectedValue)
    )
  }

  expectChecked(label: Matcher): void {
    expect(this.getView().getByLabelText(label)).toBeChecked()
  }

  expectNotChecked(label: Matcher): void {
    expect(this.getView().getByLabelText(label)).not.toBeChecked()
  }

  async changeValueByLabel(label: Matcher, newValue: unknown): Promise<void> {
    const input = this.getView().getByLabelText(label)
    await act(async () => {
      fireEvent.change(input, { target: { value: newValue } })
    })
  }

  async submitForm(): Promise<void> {
    await submitForm(this.getView().container)
  }

  async click(text: Matcher, n = 0): Promise<void> {
    const clickable = this.getView().getAllByText(text)[n]
    await act(async () => {
      fireEvent.click(clickable)
    })
  }
}
