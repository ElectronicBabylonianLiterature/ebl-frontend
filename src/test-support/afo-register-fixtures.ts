import { Factory } from 'fishery'
import Chance from 'chance'
import AfoRegisterRecord, {
  AfoRegisterRecordSuggestion,
} from 'afo-register/domain/Record'

const chance = new Chance()
const PUBLICATIONS = [
  'StOr',
  'Al.T.',
  'OECT',
  'VS',
  'STT',
  'CCT',
  'CM',
  'CST',
  'SAAB',
]

const getRandomPublication = (): string => chance.pickone(PUBLICATIONS)

const getAfoNumber = (): string => `AfO ${chance.integer({ min: 10, max: 52 })}`

const getPage = (): string => `${chance.integer({ min: 1, max: 700 })}`

const getText = (): string =>
  `${getRandomPublication()}, ${chance.integer({ min: 1, max: 40 })}`

const getTextNumber = (): string =>
  `Nr. ${chance.integer({ min: 1, max: 300 })}`

const getTextNumbers = (): string[] =>
  Array.from({ length: chance.integer({ min: 1, max: 15 }) }, getTextNumber)

const getLinesDiscussed = (): string =>
  `${chance.integer({ min: 1, max: 40 })}f.`

const getDiscussedBy = (): string =>
  `${chance.last()}, ${getRandomPublication()}, ${chance.integer({
    min: 1,
    max: 40,
  })}, ${chance.integer({ min: 1, max: 50 })}`

const getFragmentNumber = (): string[] =>
  Array.from(
    { length: chance.integer({ min: 0, max: 5 }) },
    () => `${chance.word()}.${chance.natural()}`
  )

export const afoRegisterRecordFactory = Factory.define<AfoRegisterRecord>(
  () =>
    new AfoRegisterRecord({
      afoNumber: getAfoNumber(),
      page: getPage(),
      text: getText(),
      textNumber: getTextNumber(),
      linesDiscussed: getLinesDiscussed(),
      discussedBy: getDiscussedBy(),
      discussedByNotes: chance.sentence(),
      fragmentNumbers: getFragmentNumber(),
    })
)

export const afoRegisterRecordSuggestionFactory = Factory.define<
  AfoRegisterRecordSuggestion
>(
  () =>
    new AfoRegisterRecordSuggestion({
      text: getText(),
      textNumbers: getTextNumbers().sort(),
    })
)
