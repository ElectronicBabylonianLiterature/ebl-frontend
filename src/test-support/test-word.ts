import Word from 'dictionary/domain/Word'

export const wordDto: Word = {
  _id: 'test I',
  lemma: ['test'],
  attested: true,
  legacyLemma: 'test',
  homonym: 'I',
  forms: [
    {
      lemma: [''],
      notes: ['dsdsds'],
      attested: true,
    },
    {
      lemma: ["abā'um"],
      notes: ['**BOLD**'],
      attested: true,
    },
  ],
  meaning: '\\~ "to weed" O/jB',
  amplifiedMeanings: [
    {
      meaning: '(*i/i*) field',
      vowels: [
        {
          value: ['i', 'i'],
          notes: ['dsdsds'],
        },
      ],
      key: 'G',
      entries: [
        {
          meaning: 'mmkkk',
          vowels: [
            {
              value: [''],
              notes: [],
            },
          ],
        },
      ],
    },
  ],
  logograms: [
    {
      logogram: [''],
      notes: ['dsadas'],
    },
  ],
  derived: [
    [
      {
        lemma: ['wabûtu'],
        homonym: 'I',
        notes: [],
      },
    ],
  ],
  derivedFrom: {
    lemma: ['sadasd'],
    homonym: '',
    notes: [''],
  },
  source: 'This is a test word!',
  roots: ["wb'", "'b'"],
  pos: ['V'],
  guideWord: 'weed',
  arabicGuideWord: 'weed',
  origin: 'cda',
  cdaAddenda: '(egg-shaped) bead',
  oraccWords: [],
  akkadischeGlossareUndIndices: [],
}
