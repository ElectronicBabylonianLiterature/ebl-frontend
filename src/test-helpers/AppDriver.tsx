import React from 'react'
import {
  fireEvent,
  render,
  RenderResult,
  act,
  Matcher,
  within
} from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import _ from 'lodash'
import App from 'App'
import WordRepository from 'dictionary/infrastructure/WordRepository'
import FragmentRepository from 'fragmentarium/infrastructure/FragmentRepository'
import ApiImageRepository from 'fragmentarium/infrastructure/ImageRepository'
import FragmentService from 'fragmentarium/application/FragmentService'
import WordService from 'dictionary/application/WordService'
import TextService from 'corpus/TextService'
import MemorySession, { Session, guestSession } from 'auth/Session'
import BibliographyRepository from 'bibliography/infrastructure/BibliographyRepository'
import BibliographyService from 'bibliography/application/BibliographyService'
import FragmentSearchService from 'fragmentarium/application/FragmentSearchService'
import { Promise } from 'bluebird'
import { submitForm } from 'test-helpers/utils'
import { eblNameProperty, AuthenticationContext } from 'auth/Auth'

function createApp(api): JSX.Element {
  const wordRepository = new WordRepository(api)
  const fragmentRepository = new FragmentRepository(api)
  const imageRepository = new ApiImageRepository(api)
  const bibliographyRepository = new BibliographyRepository(api)
  const wordService = new WordService(wordRepository)
  const bibliographyService = new BibliographyService(bibliographyRepository)
  const fragmentService = new FragmentService(
    fragmentRepository,
    imageRepository,
    wordRepository,
    bibliographyService
  )
  const fragmentSearchService = new FragmentSearchService(fragmentRepository)
  const textService = new TextService(api)

  return (
    <App
      wordService={wordService}
      fragmentService={fragmentService}
      fragmentSearchService={fragmentSearchService}
      bibliographyService={bibliographyService}
      textService={textService}
    />
  )
}

export default class AppDriver {
  private readonly api
  private initialEntries: string[] = []
  private element: RenderResult | null = null
  private session: Session | null = null

  constructor(api) {
    this.api = api
  }

  getElement(): RenderResult {
    if (this.element) {
      return this.element
    } else {
      throw new Error('getElement called before render.')
    }
  }

  withPath(path): AppDriver {
    this.initialEntries = [path]
    return this
  }

  withSession(): AppDriver {
    this.session = new MemorySession([
      'write:texts',
      'read:fragments',
      'annotate:fragments',
      'read:words'
    ])
    return this
  }

  async render(): Promise<AppDriver> {
    await act(async () => {
      this.element = render(
        <MemoryRouter initialEntries={this.initialEntries}>
          <AuthenticationContext.Provider
            value={{
              login: _.noop,
              logout: _.noop,
              getSession: (): Session => this.session ?? guestSession,
              isAuthenticated: (): boolean => this.session !== null,
              getAccessToken(): Promise<string> {
                throw new Error('Not implemented')
              },
              getUser(): { [eblNameProperty]: string } {
                return { [eblNameProperty]: 'Test' }
              }
            }}
          >
            {createApp(this.api)}
          </AuthenticationContext.Provider>
        </MemoryRouter>
      )
    })

    return this
  }

  async waitForText(text: Matcher): Promise<void> {
    await this.getElement().findByText(text)
  }

  expectTextContent(text: string | RegExp): void {
    expect(this.getElement().container).toHaveTextContent(text)
  }

  expectNotInContent(text: Matcher): void {
    expect(this.getElement().queryByText(text)).toBeNull()
  }

  expectLink(text: Matcher, expectedHref: string): void {
    expect(this.getElement().getByText(text)).toHaveAttribute(
      'href',
      expectedHref
    )
  }

  expectBreadcrumbs(crumbs: readonly string[]): void {
    this.expectTextContent(crumbs.join(''))
  }

  expectBreadcrumb(crumb: string, link: string): void {
    expect(
      within(this.getElement().getByLabelText('breadcrumb')).getByText(crumb)
    ).toHaveAttribute('href', link)
  }

  expectInputElement(label: Matcher, expectedValue: unknown): void {
    expect(
      (this.getElement().getByLabelText(label) as HTMLInputElement).value
    ).toEqual(String(expectedValue))
  }

  changeValueByLabel(label: Matcher, newValue): void {
    const input = this.getElement().getByLabelText(label)
    act(() => {
      fireEvent.change(input, { target: { value: newValue } })
    })
  }

  async submitForm(): Promise<void> {
    await submitForm(this.getElement())
  }

  async click(text: Matcher, n = 0): Promise<void> {
    const clickable = this.getElement().getAllByText(text)[n]
    await act(async () => {
      fireEvent.click(clickable)
    })
  }
}
