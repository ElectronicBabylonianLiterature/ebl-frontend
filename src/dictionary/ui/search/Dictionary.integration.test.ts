import FakeApi from 'test-support/FakeApi'
import AppDriver from 'test-support/AppDriver'
import { wordDto } from 'test-support/test-word'

const query = 'lem[ma?]'
const words = [wordDto]

let fakeApi: FakeApi
let appDriver: AppDriver

afterEach(() => {
  fakeApi.verifyExpectations()
})

beforeEach(async () => {
  fakeApi = new FakeApi()
  appDriver = await new AppDriver(fakeApi.client)
    .withSession()
    .withPath(`/dictionary`)
    .render()
})

test('Snapshot', () => {
  expect(appDriver.getElement().container).toMatchSnapshot()
})

test('Query', async () => {
  fakeApi.expectSearchWords(query, words)
  await appDriver.changeValueByLabel('Query', query)
  await appDriver.submitForm()
  await appDriver.waitForText(words[0].lemma.join(' '))
  expect(appDriver.getElement().container).toMatchSnapshot()
})
