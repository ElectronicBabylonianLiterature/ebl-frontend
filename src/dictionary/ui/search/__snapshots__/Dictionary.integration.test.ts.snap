// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query 1`] = `
<div>
  <header
    class="Header"
  >
    <nav
      class="navbar navbar-expand-md navbar-light"
    >
      <div
        class="container"
      >
        <a
          class="active navbar-brand"
          href="/"
          title="electronic Babylonian Literature (eBL)"
        >
          <h1
            class="Header__title"
          >
            <span
              class="Header__title-main"
            >
              electronic
              <br />
              Babylonian
              <br />
              Literature
            </span>
            <small
              class="Header__title-abbreviation"
            >
              eBL
            </small>
          </h1>
        </a>
        <div
          class="navbar-collapse collapse"
          id="Header-1"
        >
          <div
            class="mx-auto navbar-nav"
          >
            <div
              class="nav-item"
            >
              <a
                class="active nav-link"
                data-rb-event-key="/dictionary"
                href="/dictionary"
              >
                Dictionary
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                class="nav-link"
                data-rb-event-key="/corpus"
                href="/corpus"
              >
                Corpus
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                class="nav-link"
                data-rb-event-key="/fragmentarium"
                href="/fragmentarium"
              >
                Fragmentarium
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                class="nav-link"
                data-rb-event-key="/bibliography"
                href="/bibliography"
              >
                Bibliography
              </a>
            </div>
          </div>
          <span
            class="navbar-text"
          >
            <button
              class="btn btn-outline-secondary btn-sm"
              type="button"
            >
              Logout 
              Test
            </button>
          </span>
        </div>
        <button
          aria-controls="Header-1"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
      </div>
    </nav>
  </header>
  <section
    class="App-content"
  >
    <header
      class="App-content__header"
    >
      <nav
        aria-label="breadcrumb"
      >
        <ol
          class="breadcrumb"
        >
          <li
            class="breadcrumb-item active"
          >
            <a
              href="/"
            >
              eBL
            </a>
          </li>
          <li
            aria-current="page"
            class="breadcrumb-item active"
          >
            Dictionary
          </li>
        </ol>
      </nav>
      <div
        class="float-right btn-toolbar"
        role="toolbar"
      />
      <h2>
        Dictionary
      </h2>
    </header>
    <div
      class="Dictionary-search"
    >
      <form
        class=""
      >
        <div
          class="form-group row"
        >
          <label
            class="form-label col-form-label col-sm-2"
            for="query"
          >
            Query
          </label>
          <div
            class="col-sm-6"
          >
            <input
              class="form-control"
              id="query"
              placeholder="lemma or meaning"
              type="text"
              value="lem[ma?]"
            />
          </div>
          <div
            class="col-sm-4"
          >
            <button
              class="btn btn-primary"
              type="submit"
            >
              Query
            </button>
          </div>
        </div>
      </form>
    </div>
    <ul
      class="WordSearch-results"
    >
      <li
        class="WordSearch-results__result"
      >
        <div
          class="Word"
        >
          <a
            class="BibliographySearch__edit"
            href="/dictionary/test I/edit"
          >
            <i
              class="fas fa-edit"
            />
          </a>
          <dfn
            title="test I"
          >
            <strong>
              <a
                href="/dictionary/test I"
              >
                
                test
              </a>
            </strong>
             I
          </dfn>
          <ul
            class="Word__forms"
          >
            <li
              class="Word__form"
            >
              <span
                class="Notes-note"
              >
                dsdsds
                 
              </span>
              <em />
            </li>
            <li
              class="Word__form"
            >
              <span
                class="Notes-note"
              >
                <strong>
                  BOLD
                </strong>
                 
              </span>
              <em>
                abā'um
              </em>
            </li>
          </ul>
           
          ~
           "to weed" O/jB
           
          <ul>
            <li>
              <strong>
                G
              </strong>
               
              (
              <em>
                i/i
              </em>
              ) field
               
              <ul>
                <li
                  class="AmplifiedMeanings__entry"
                >
                  <strong>
                    1.
                  </strong>
                   
                  mmkkk
                </li>
              </ul>
            </li>
          </ul>
           
          <ul>
            <li>
              <ul
                class="Derived__group"
              >
                <li
                  class="Derived__entry"
                >
                  <em>
                    wabûtu
                  </em>
                   I
                </li>
              </ul>
            </li>
          </ul>
           
          <span
            class="Word__derivedFrom"
          >
            <em>
              sadasd
            </em>
            
          </span>
        </div>
      </li>
    </ul>
  </section>
</div>
`;

exports[`Snapshot 1`] = `
<div>
  <header
    class="Header"
  >
    <nav
      class="navbar navbar-expand-md navbar-light"
    >
      <div
        class="container"
      >
        <a
          class="active navbar-brand"
          href="/"
          title="electronic Babylonian Literature (eBL)"
        >
          <h1
            class="Header__title"
          >
            <span
              class="Header__title-main"
            >
              electronic
              <br />
              Babylonian
              <br />
              Literature
            </span>
            <small
              class="Header__title-abbreviation"
            >
              eBL
            </small>
          </h1>
        </a>
        <div
          class="navbar-collapse collapse"
          id="Header-1"
        >
          <div
            class="mx-auto navbar-nav"
          >
            <div
              class="nav-item"
            >
              <a
                class="active nav-link"
                data-rb-event-key="/dictionary"
                href="/dictionary"
              >
                Dictionary
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                class="nav-link"
                data-rb-event-key="/corpus"
                href="/corpus"
              >
                Corpus
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                class="nav-link"
                data-rb-event-key="/fragmentarium"
                href="/fragmentarium"
              >
                Fragmentarium
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                class="nav-link"
                data-rb-event-key="/bibliography"
                href="/bibliography"
              >
                Bibliography
              </a>
            </div>
          </div>
          <span
            class="navbar-text"
          >
            <button
              class="btn btn-outline-secondary btn-sm"
              type="button"
            >
              Logout 
              Test
            </button>
          </span>
        </div>
        <button
          aria-controls="Header-1"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
      </div>
    </nav>
  </header>
  <section
    class="App-content"
  >
    <header
      class="App-content__header"
    >
      <nav
        aria-label="breadcrumb"
      >
        <ol
          class="breadcrumb"
        >
          <li
            class="breadcrumb-item active"
          >
            <a
              href="/"
            >
              eBL
            </a>
          </li>
          <li
            aria-current="page"
            class="breadcrumb-item active"
          >
            Dictionary
          </li>
        </ol>
      </nav>
      <div
        class="float-right btn-toolbar"
        role="toolbar"
      />
      <h2>
        Dictionary
      </h2>
    </header>
    <div
      class="Dictionary-search"
    >
      <form
        class=""
      >
        <div
          class="form-group row"
        >
          <label
            class="form-label col-form-label col-sm-2"
            for="query"
          >
            Query
          </label>
          <div
            class="col-sm-6"
          >
            <input
              class="form-control"
              id="query"
              placeholder="lemma or meaning"
              type="text"
              value=""
            />
          </div>
          <div
            class="col-sm-4"
          >
            <button
              class="btn btn-primary"
              type="submit"
            >
              Query
            </button>
          </div>
        </div>
      </form>
    </div>
    <ul
      class="WordSearch-results"
    />
  </section>
</div>
`;
